File: 94. Binary Tree Inorder Traversal/94-recursive.js
// 94. Binary Tree Inorder Traversal recursive solution in JavaScript
// https://leetcode.com/problems/binary-tree-inorder-traversal/description/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 var inorderTraversal = function(root) {
    if (!root) {
        return [];
    }

    return [
        ...inorderTraversal(root.left),
        root.val,
        ...inorderTraversal(root.right)
    ]
};

File: 94. Binary Tree Inorder Traversal/94-recursive.go
// 94. Binary Tree Inorder Traversal recursive solution in GoLang
// https://leetcode.com/problems/binary-tree-inorder-traversal/description/

package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	var output []int
	traverse(root, &output)

	return output
}

func traverse(root *TreeNode, output *[]int) {
	if root == nil {
		return
	}

	traverse(root.Left, output)
	*output = append(*output, root.Val)
	traverse(root.Right, output)
}
